### PR TITLE
New Post: Building a sidenav component

### LINKED ISSUE #3565: content: A Sidenav Component [2020-09-03]
* [draft](https://docs.google.com/document/d/1QJWdvxOwl-V1B7U9_-ZLD86L5gylLlEmDCPTPOBgWts)
* [thread](https://groups.google.com/a/google.com/g/webdev-content/c/se_huYW36R0)
* next step: @kaycebasques to review

### PR DESCRIPTION
**Pull request by argyleink**
Fixes #3565 

TODO:
- [x] get editorial reviews
- [x] link to YouTube in post and codelab
- [ ] release with the associated video
- [x] set the date in the blog post

### CODE REVIEW

**Comment by argyleink** [comment]
no deploy preview for draft PRs?

File: src/site/content/en/blog/building-a-sidenav-component/index.md [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+
+Our layout is now stacked at a mobile viewport size. Until I add some new styles, 
+it's overlaying our article by default. Here's some UX I'm shooting for in this next section:
+
+- Animate open and close
+- Only animate with motion if the user is OK with that
+- Animate `visibility` so keyboard focus doesn't enter the offscreen element
+
+Animate between open and close, `:target` and `:not(:target)`, 
+did someone say animation or motion!? Let's start with accessibility.
```
Comment (kaycebasques) [inline src/site/content/en/blog/building-a-sidenav-component/index.md:169]: Can you rephrase this? I'm not sure what you mean here.

File: src/site/content/en/blog/building-a-sidenav-component/index.md [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+
+The `Escape` key on your keyboard should close the menu right? Let's wire that up.
+
+```js
+const sidenav = document.querySelector('#sidenav-open')
+
+sidenav.addEventListener('keyup', event => {
+  if (event.code === 'Escape')
+    document.location.hash = ''
+})
```
Comment (kaycebasques) [inline src/site/content/en/blog/building-a-sidenav-component/index.md:300]:
```suggestion
const sidenav = document.querySelector('#sidenav-open');

sidenav.addEventListener('keyup', event => {
  if (event.code === 'Escape') document.location.hash = '';
});
```

* https://google.github.io/styleguide/jsguide.html#formatting-braces-all
* https://google.github.io/styleguide/jsguide.html#formatting-semicolons-are-required

Can you update the JS in your Glitch to also follow the style guide? I know that these things can turn into debates but the fact is we're on a Google site, publishing code associated to Google's brand, so we should follow the Google JS style guide.

File: src/site/content/en/blog/building-a-sidenav-component/index.md [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+they open or close. I want to make toggling easy. 
+
+```js
+sidenav.addEventListener('transitionend', e => {
+  const isOpen = document.location.hash === '#sidenav-open'
+
+  isOpen
+    ? document.querySelector('#sidenav-close').focus()
+    : document.querySelector('#sidenav-button').focus()
+})
```
Comment (kaycebasques) [inline src/site/content/en/blog/building-a-sidenav-component/index.md:315]:
```suggestion
  const isOpen = document.location.hash === '#sidenav-open';

  isOpen
      ? document.querySelector('#sidenav-close').focus()
      : document.querySelector('#sidenav-button').focus();
})
```
https://google.github.io/styleguide/jsguide.html#formatting-indent

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md [hunk 4]
```diff
@@ -0,0 +1,332 @@
+---
+layout: codelab
+title: "Codelab: Building a Sidenav component"
+authors:
+  - adamargyle
+description: |
+  Learn how to build a responsive slide out side navigation layout component.
+date: 2020-11-25
+hero: hero.jpg
+thumbnail: thumb.jpg
```
Comment (kaycebasques) [inline src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:10]:
```suggestion
```
Hero and thumbnail pages have no effect when we use codelab layout, right? Please remove the hero image from this directory.

File: src/site/content/en/blog/building-a-sidenav-component/index.md [hunk 5]
```diff
@@ -0,0 +1,333 @@
+---
+layout: post
+title: Building a sidenav component
+subhead: A foundational overview of how to build a responsive slide out sidenav
+authors:
+  - adamargyle
+description: A foundational overview of how to build a responsive slide out sidenav
+date: 2021-01-16
+hero: hero.png
```
Comment (kaycebasques) [inline src/site/content/en/blog/building-a-sidenav-component/index.md:9]: Can you run this hero through an optimizer and try to get its size down a bit? It doesn't have to be an aggressive optimization. Just trying to keep the size of the repo down.

File: src/site/content/en/blog/building-a-sidenav-component/index.md [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+<figure class="w-figure w-figure--fullbleed">
+  <video playsinline controls autoplay loop muted class="w-screenshot">
+    <source src="https://storage.googleapis.com/atoms-sandbox.google.com.a.appspot.com/gui-sidenav/mobile-demo-1080p.mp4">
+  </video>
+  <figcaption class="w-figure">
+    Light and dark theme down on iOS and Android
+  </figcaption>
+</figure>
+
+## Picking tools {: #tools }
```
Comment (kaycebasques) [inline src/site/content/en/blog/building-a-sidenav-component/index.md:54]: Tools might not be the best word here. I think some readers might assume that you mean DevTools or build tools.

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+  - mobile
+  - ux
+---
+
+This codelab teaches you how to build a responsive slide out side navigation layout component
+on the web. We'll build the component as we go, starting with HTML, then CSS,
+then JavaScript.
+
+Check out my blog post [Building a Sidenav component](/building-a-sidenav-component)
+to learn about the CSS tools chosen for building this component.
```
Comment (kaycebasques) [inline src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:28]: As mentioned in the blog post, I worry "tools" might be a bit misleading for some readers.

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+## HTML
+
+First, get the essentials of the HTML setup so there's content and some boxes to work with.
+I'll be using an `<aside>` to hold the navigation menu, and a `<main>` to hold the primary page content. 
+Drop the following HTML into the body tag.
+
+```html
+<aside></aside>
+<main></main>
+```
```
Comment (kaycebasques) [inline src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:46]: Remember that codelabs should be action-focused. We want them to learn by doing. We should provide instructions first and foremost, and then context.

```
Drop the following HTML into the body tag.

    …
    <body>
      <aside></aside>
      <main></main>
    </body>
    …

The `<aside>` holds the navigation menu, and the `<main>` holds the primary page content. 
```

* Please review your insructions throughout the codelab and check if you can refactor to this pattern of "instruction first, and then context after the code sample".
* We should provide a little more context in the code samples so that there is no ambiguity about where the reader should put the code. Please review all of your code samples to see if they need a bit more context. (This one is obvious but in general it can quickly become unobvious as you progress through a codelab.)
* Please replace `I` and `I'll` with `you` throughout the post. The focus here is that the reader should do the action. Not us. For this reason we want to make it clear that **you** need to do X.

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+
+Next I'll fill in those semantic elements with the rest of the page content. Add a 
+navigation element, some nav links and a close link.
```
Comment (kaycebasques) [inline src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:49]:
```suggestion
navigation element, some nav links and a close link inside the `<aside>` element
that you just added to the page.
```

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+    <a href="#">Admin</a>
+  </nav>
+  
+  <a href="#"></a>
+</aside>
+```
+
+In the main content element, I added a header and an article to semantically hold the layout content. 
+The header will have the menu open button, we'll show and hide it based on viewport size soon. The aside 
+has the close button.
```
Comment (kaycebasques) [inline src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:72]: We're missing a clear instruction here. You mention what you did but it's not immediately clear what the reader has to do. Of course I know that you mean that they should add this header and article element to the main element. But we shouldn't make our readers have to do a mental diff in their head to figure out what they have to do.

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md (hunk 8 above)
Comment (kaycebasques) [reply src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:46]: (Meta note: please leave this convo unresolved so I can remember to check back on these points throughout the doc later)

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+  <article>
+    {put some placeholder content here}
+  </article>
+</main>
+```
+
+So far I've added an aside element, with a nav, links and a way to close the sidenav. 
+Also added a header, a way to open the sidenav, and an article to the main element. 
+This is clean, semantic and pretty timeless already, but next I want to 
+go sprinkle on some accessibility sugar, to make it clean and clear for everyone.
```
Comment (kaycebasques) [inline src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:96]: This is somewhat nitpicky but I don't think we should call accessibility `sugar` because it's a fundamental requirement for people who rely on screen readers etc. `sugar` reminds me of `syntactic sugar` which implies a "nice-to-have" thing rather than a table-stakes requirement.

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+```html//0
+<a href="#"></a>
+<a href="#" id="sidenav-close" title="Close Menu" aria-label="Close Menu"></a>
+```
+
+## CSS
+
+Time to layout the elements. The main content 
+and sidenav are inside the `<body>` tag, so I started with their layout. Add the following 
+CSS into `css/sidenav.css` so the body element positions the children.
```
Comment (kaycebasques) [inline src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:123]: The last sentence here is a good example of how we should phrase instructions to our readers.

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+    
+    &:target {
+      visibility: visible;
+    }
+  }
+}
+```
+
+I also only need the open and close buttons when at a mobile layout, let's 
+manage them with CSS and a media query as well.
```
Comment (kaycebasques) [inline src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:223]: I'll just give another example so you have more context about what I'm looking for with my requested changes. In this case the text should be rephased to be a straight-up instruction:

    Add CSS and a media query so that you can manage the open and close buttons when at a mobile layout

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md [hunk 14]
```diff
@@ -0,0 +1,332 @@
...
+    <source src="https://storage.googleapis.com/atoms-sandbox.google.com.a.appspot.com/prefers-reduced-motion.mp4">
+  </video>
+  <figcaption class="w-figure">
+    demo of the interaction with and without duration applied based on `prefers-reduced-motion` media query
+  </figcaption>
+</figure>
+
+### Sprinkle in some Javascript
+
+The `esc` key should close the menu. I add this by listening for a key event on the sidenav element,
```
Comment (kaycebasques) [inline src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:281]: In the blog post I believe you spelled out `Escape`. Consistency is good!

**Review by kaycebasques — CHANGES_REQUESTED** [review]
Looking good. The main thing that I want you to focus on is rephrasing the codelab to shift from what you did to what the reader needs to do.

In other words, before each code block, I'd like to see you rephrase it to a very straightforward instruction for the reader.

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md (hunk 4 above)
Comment (argyleink) [reply src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:10]: here's why i add them, which reminds me, the centering post needs the fix too.

image

File: src/site/content/en/blog/building-a-sidenav-component/index.md (hunk 1 above)
Comment (argyleink) [reply src/site/content/en/blog/building-a-sidenav-component/index.md:169]: replaced with:

`As I begin to implement motion animations, I want to start with accessibility top of mind.`

File: src/site/content/en/blog/building-a-sidenav-component/index.md (hunk 2 above)
Comment (argyleink) [reply src/site/content/en/blog/building-a-sidenav-component/index.md:300]: np

File: src/site/content/en/blog/building-a-sidenav-component/index.md (hunk 6 above)
Comment (argyleink) [reply src/site/content/en/blog/building-a-sidenav-component/index.md:54]: nice, i agree. i've replaced usages of it with either tactics or web platform features. let me know what you think 👍🏻

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md (hunk 11 above)
Comment (argyleink) [reply src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:97]: fair! updated

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md (hunk 12 above)
Comment (argyleink) [reply src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:124]: i like the updates that use more of this verbiage, great call 👍🏻

**Comment by argyleink** [comment]
all feedback has been implemented! the codelab got a big boost, thanks 👍🏻 

ready for another pass 🙂

**Review by kaycebasques — CHANGES_REQUESTED** [review]
@argyleink I've got various copyedits in this review.

The bigger thing we need to sort out, however, is that I just tried to usability test the codelab and it's not working as intended. It very well could have been pilot error. But I'll need to walk through it again on Monday and figure out where I got stuck.

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md [hunk 15]
```diff
@@ -0,0 +1,379 @@
...
+```
+
+## Try it out
+
+{% Instruction 'preview' %}
+
+## Conclusion
+
+That's a wrap up for the needs I had with the component. Feel free to build upon
+it, drive it with data, and in general make it yours! There's always more to add
```
[the remark this answers is not in the document]
Comment (argyleink) [reply src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:374]: updated to be more specific 👍🏻 

"Feel free to build upon it, drive it with JavaScript state instead of the URL, and in general make it yours!"

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md [hunk 16]
```diff
@@ -0,0 +1,379 @@
...
+      visibility: visible;
+      transform: translateX(0);
+      transition: transform var(--duration) var(--easeOutExpo);
+    }
+  }
+
+  @media (prefers-reduced-motion: reduce) {
+    --duration: 1ms;
+  }
+}
```
[the remark this answers is not in the document]
Comment (argyleink) [reply src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:313]: awesome! that's super handy to know, ty. will update 👍🏻

File: src/site/content/en/blog/codelab-building-a-sidenav-component/index.md [hunk 17]
```diff
@@ -0,0 +1,379 @@
...
+    overflow: hidden auto;
+    overscroll-behavior: contain;
+
+    visibility: hidden; /* not keyboard accessible when closed */
+  }
+}
+```
+
+Go ahead and open the side menu after scrolling the page, 
+or try scrolling the page while the sidenav is open. What do you think?
```
[the remark this answers is not in the document]
Comment (argyleink) [reply src/site/content/en/blog/codelab-building-a-sidenav-component/index.md:249]: moved it a section down 1, thanks!

**Comment by argyleink** [comment]
aight @kaycebasques, back to you! 
- [x] codelab completes when following instruction
- [x] all comments and suggestions were added or resolved

**Comment by robdodson** [comment]
Adam asked if we could deploy this as the video is going live in an hour.
